Fail to reset-cursor on partitioned topic

After issues with ZK #308 we are attempting to reset some consumers and we're getting the following exceptions.
How and why can a `ManagedCursorImpl` have a `ReadOnlyLedgerHandle`?

```
Caused by: com.yahoo.pulsar.broker.service.BrokerServiceException$SubscriptionInvalidCursorPosition: unable to persist position for cursor reset 730778:44020
	at com.yahoo.pulsar.broker.service.persistent.PersistentSubscription$5$1.resetFailed(PersistentSubscription.java:371)
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl$7.operationFailed(ManagedCursorImpl.java:746)
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl.lambda$persistPosition$12(ManagedCursorImpl.java:1911)
	at org.apache.bookkeeper.client.ReadOnlyLedgerHandle.asyncAddEntry(ReadOnlyLedgerHandle.java:105)
	at org.apache.bookkeeper.client.ReadOnlyLedgerHandle.asyncAddEntry(ReadOnlyLedgerHandle.java:98)
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl.persistPosition(ManagedCursorImpl.java:1889)
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl.internalResetCursor(ManagedCursorImpl.java:768)
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl.lambda$asyncResetCursor$4(ManagedCursorImpl.java:780)
	at org.apache.bookkeeper.mledger.util.SafeRun$1.safeRun(SafeRun.java:27)
	at org.apache.bookkeeper.util.SafeRunnable.run(SafeRunnable.java:31)
```

--- a/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java
+++ b/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java
@@ -240,7 +240,6 @@ public class ManagedCursorImpl implements ManagedCursor {
             }
 
             // Read the last entry in the ledger
-            cursorLedger = lh;
             lh.asyncReadLastEntry((rc1, lh1, seq, ctx1) -> {
                 if (log.isDebugEnabled()) {
                     log.debug("readComplete rc={} entryId={}", rc1, lh1.getLastAddConfirmed());
@@ -762,24 +761,17 @@ public class ManagedCursorImpl implements ManagedCursor {
 
         };
 
-        if (cursorLedger == null) {
-            persistPositionMetaStore(-1, newPosition, new MetaStoreCallback<Void>() {
-                @Override
-                public void operationComplete(Void result, Stat stat) {
-                    log.info("[{}] Updated cursor {} with ledger id {} md-position={} rd-position={}",
-                            ledger.getName(), name, -1, markDeletePosition, readPosition);
-                    cursorLedgerStat = stat;
-                    finalCallback.operationComplete();
-                }
+        internalAsyncMarkDelete(newPosition, new MarkDeleteCallback() {
+            @Override
+            public void markDeleteComplete(Object ctx) {
+                finalCallback.operationComplete();
+            }
 
-                @Override
-                public void operationFailed(MetaStoreException e) {
-                    finalCallback.operationFailed(e);
-                }
-            });
-        } else {
-            persistPosition(cursorLedger, newPosition, finalCallback);
-        }
+            @Override
+            public void markDeleteFailed(ManagedLedgerException exception, Object ctx) {
+                finalCallback.operationFailed(exception);
+            }
+        }, null);
     }
 
     @Override
@@ -1817,19 +1809,8 @@ public class ManagedCursorImpl implements ManagedCursor {
 
     private void flushPendingMarkDeletes() {
         if (!pendingMarkDeleteOps.isEmpty()) {
-            if (RESET_CURSOR_IN_PROGRESS_UPDATER.get(this) == TRUE) {
-                failPendingMarkDeletes();
-            } else {
-                internalFlushPendingMarkDeletes();
-            }
+            internalFlushPendingMarkDeletes();
         }
-    }
-
-    private void failPendingMarkDeletes() {
-        for (PendingMarkDeleteEntry e : pendingMarkDeleteOps) {
-            e.callback.markDeleteFailed(new ManagedLedgerException("reset cursor in progress"), e.ctx);
-        }
-        pendingMarkDeleteOps.clear();
     }
 
     void internalFlushPendingMarkDeletes() {

--- a/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerBkTest.java
+++ b/managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerBkTest.java
@@ -426,4 +426,36 @@ public class ManagedLedgerBkTest extends BookKeeperClusterTestCase {
         factory.shutdown();
         assertNotNull(offlineTopicStats);
     }
+
+    @Test(timeOut = 20000)
+    void testResetCursorAfterRecovery() throws Exception {
+        ManagedLedgerFactory factory = new ManagedLedgerFactoryImpl(bkc, zkc);
+        ManagedLedgerConfig conf = new ManagedLedgerConfig().setMaxEntriesPerLedger(10).setEnsembleSize(1)
+                .setWriteQuorumSize(1).setAckQuorumSize(1).setMetadataEnsembleSize(1).setMetadataWriteQuorumSize(1)
+                .setMetadataAckQuorumSize(1);
+        ManagedLedger ledger = factory.open("my_test_move_cursor_ledger", conf);
+        ManagedCursor cursor = ledger.openCursor("trc1");
+        Position p1 = ledger.addEntry("dummy-entry-1".getBytes());
+        Position p2 = ledger.addEntry("dummy-entry-2".getBytes());
+        Position p3 = ledger.addEntry("dummy-entry-3".getBytes());
+        Position p4 = ledger.addEntry("dummy-entry-4".getBytes());
+
+        cursor.markDelete(p3);
+
+        ManagedLedgerFactory factory2 = new ManagedLedgerFactoryImpl(bkc, zkc);
+        ledger = factory2.open("my_test_move_cursor_ledger", conf);
+        cursor = ledger.openCursor("trc1");
+
+        assertEquals(cursor.getMarkDeletedPosition(), p3);
+        assertEquals(cursor.getReadPosition(), p4);
+        assertEquals(cursor.getNumberOfEntriesInBacklog(), 1);
+
+        cursor.resetCursor(p2);
+        assertEquals(cursor.getMarkDeletedPosition(), p1);
+        assertEquals(cursor.getReadPosition(), p2);
+        assertEquals(cursor.getNumberOfEntriesInBacklog(), 3);
+
+        factory2.shutdown();
+        factory.shutdown();
+    }
 }

